### PR TITLE
[opentitantool] Support inverted GPIO sense

### PR DESCRIPTION
**Pull request by jesultra**
Introduce GpioPinWrapper to handle optional inversion of GPIO sense.

### CODE REVIEW

**Review by jwnrt — APPROVED** [review]
Thanks @jesultra, this code looks correct to me though I haven't tested it.

I do want to register that I think these wrappers make opentitantool/lib even harder to work on, though. Do you see any way of integrating the `invert` setting into the existing GPIO code?

File: sw/host/opentitanlib/src/app/mod.rs [hunk 1]
```diff
@@ -653,10 +678,28 @@ impl TransportWrapper {
...
+                pin.clone()
+            } else {
+                self.transport.gpio_pin(resolved_pin_name.as_str())?
+            };
+            let mut invert = false;
+            if let Some(conf) = self.pin_conf_map.get(&resolved_pin_name) {
+                if let Some(inv) = conf.invert {
+                    invert = inv;
+                }
+            }
```
Comment (engdoreis) [inline sw/host/opentitanlib/src/app/mod.rs:697]:
```suggestion
           let invert = self.pin_conf_map.get(&resolved_pin_name).unwrap_or(false).unwrap_or(false);
```

**Review by engdoreis — APPROVED** [review]

File: sw/host/opentitanlib/src/app/mod.rs (hunk 1 above)
Comment (nbdd0121) [reply sw/host/opentitanlib/src/app/mod.rs:697]: The suggestion wouldn't type check. If should be 
```rust
let invert = self.pin_conf_map.get(&resolved_pin_name).and_then(|conf| conf.invert).unwrap_or(false);
```

Comment (jesultra) [reply sw/host/opentitanlib/src/app/mod.rs:697]: Done

**Comment by jwnrt** [comment]
> Douglas has the PR #19519, which introduced an inversion feature in the chip_whisperer transport backend, and thereby avoided adding a layer. That however, left ultradebug, hyperdebug, dediprog and other transport backends without support for inverting GPIO pins, and adding the support would mean replicating the same code five places, as well as put additional burden on any future backend transport driver. I strongly wanted to avoid that.

That sounds reasonable to me, thanks @jesultra

> What exactly do you mean "harder to work on". As in harder for the code that uses opentitanlib to perform testing operations, or harder to make enhancements in opentitanlib itself.

Yes, I meant harder to maintain and develop opentitanlib itself.

I should have qualified this by saying my personal experience is that making changes to opentitanlib can be quite tricky because there are quite a few layers of indirection and abstraction (e.g. through `&dyn Any`, `TransportWrapper`, and the IO traits) with structures passing between the layers (primarily the CLI option structs). It can be hard to understand the overall control flow of an action and know where to make changes.

I think going forward `opentitanlib` would benefit from something like an `ARCHITECTURE.md` file describing all the moving parts, or if possible some breaking up into a few libraries that are less tightly coupled (in particular, I think the `test_utils`, `dif`, `transport`, and CLI parts should be separated).

**Pull request merged**